File: crypto.go
// History: Oct 02 13 tcolar Creation

package drupalcrypto

import (
  "encoding/base64"
  "errors"
  "log"
  "math"
  "strings"
)

type PhpCrypto struct {
  Scramble1 string
  Scramble2 string
  Adj       float64
  Mod       int
}

// Get a Crypto instance initialized as it is in Drupal / Ubbercart
func DrupalCrypto() (crypto PhpCrypto) {
  return PhpCrypto{
    Scramble1: "! #$%&()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[]^_`\"abcdefghijklmnopqrstuvwxyz{|}~",
    Scramble2: "f^jAE]okIOzU[2&q1{3`h5w_794p@6s8?BgP>dFV=m\" D<TcS%Ze|r:lGK/uCy.Jx)HiQ!#$~(;Lt-R}Ma,NvW+Ynb*0X",
    Adj:       1.75,
    Mod:       3,
  }
}

// This is how some Drupal data is encrypted (convert to b64 first)
func (crypto PhpCrypto) B64Crypt(data string, encKey string) (result string, err error) {
  // First to base64
  bytes := []byte(data)
  b64 := base64.StdEncoding.EncodeToString(bytes)
  return crypto.Crypt(b64, encKey)
}

func (crypto PhpCrypto) Crypt(data string, encKey string) (result string, err error) {

  var fudgefactor []float64
  fudgefactor, err = crypto.convertKey(encKey)
  if err != nil {
    return result, err
  }

  var factor2 float64

  for _, c := range data {
    c1 := string(c)
    num1 := strings.Index(crypto.Scramble1, c1)
    if num1 < 0 {
      return result, errors.New("Source string contains an invalid character : " + c1)
    }
    //log.Printf("%d num1", num1)

    adj := crypto.applyFudgeFactor(&fudgefactor)
    factor1 := factor2 + adj
    //log.Printf("%f factor1", factor1)
    num2 := int(Rounded(factor1, 0)) + num1
    num2 = crypto.checkRange(num2)
    //log.Printf("%d num2", num2)
    factor2 = factor1 + float64(num2)

    char2 := crypto.Scramble2[num2 : num2+1]
    result += char2
  }

  return result, err
}

// This is how some Drupal data is encrypted
func (crypto PhpCrypto) B64Decrypt(data string, encKey string) (result string, err error) {
  str, err := crypto.Decrypt(data, encKey)
  if err != nil {
    return result, err
  }
  log.Print(str)
  decoded, err := base64.StdEncoding.DecodeString(str)
  return string(decoded), err
}

func (crypto PhpCrypto) Decrypt(data string, encKey string) (result string, err error) {
  // Convert key into sequence of numbers
  if len(data) == 0 {
    return result, errors.New("No data supplied.")
  }

  var fudgefactor []float64
  fudgefactor, err = crypto.convertKey(encKey)
  if err != nil {
    return result, err
  }

  var factor2 float64

  for _, c := range data {
    c2 := string(c)
    num2 := strings.Index(crypto.Scramble2, c2)
    if num2 < 0 {
      return result, errors.New("Key contains an invalid character : " + c2)
    }

    adj := crypto.applyFudgeFactor(&fudgefactor)
    factor1 := factor2 + adj
    num1 := num2 - int(Rounded(factor1, 0))
    num1 = crypto.checkRange(num1)
    factor2 = factor1 + float64(num2)

    char1 := crypto.Scramble1[num1 : num1+1]
    result += char1
  }

  return result, err
}

func (crypto PhpCrypto) convertKey(key string) (result []float64, err error) {
  result = append(result, float64(len(key)))
  tot := 0.0

  for _, c := range key {
    num := strings.IndexRune(crypto.Scramble1, c)
    if num < 0 {
      return result, errors.New("Key contains an invalid character : " + string(c))
    }
    result = append(result, float64(num))
    tot += float64(num)
  }
  result = append(result, tot)

  return result, err
}

func (crypto PhpCrypto) applyFudgeFactor(fudgefactor *[]float64) (fudge float64) {
  f := *fudgefactor
  fudge = f[0]         // take first element
  *fudgefactor = f[1:] // remove it from array

  fudge += crypto.Adj
  *fudgefactor = append(*fudgefactor, fudge)

  if int(fudge)%crypto.Mod == 0 {
    fudge *= -1
  }

  return fudge
}

func (crypto PhpCrypto) checkRange(input int) (num int) {
  num = input
  //num = Rounded(num, 0)
  limit := len(crypto.Scramble1)

  for num >= limit {
    num = num - limit
  }
  for num < 0 {
    num = num + limit
  }
  return num
}

func Rounded(val float64, decimals int) (rounded float64) {
  var rounder float64
  pow := math.Pow(10, float64(decimals))
  intermed := val * pow
  _, frac := math.Modf(intermed)
  if frac >= 0.5 {
    rounder = math.Ceil(intermed)
  } else {
    rounder = math.Floor(intermed)
  }
  return rounder / pow
}

File: crypto.php
<?php

class uc_encryption_class {

  protected static $scramble1 = '! #$%&()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[]^_`"abcdefghijklmnopqrstuvwxyz{|}~';
  protected static $scramble2 = 'f^jAE]okIOzU[2&q1{3`h5w_794p@6s8?BgP>dFV=m" D<TcS%Ze|r:lGK/uCy.Jx)HiQ!#$~(;Lt-R}Ma,NvW+Ynb*0X';

  protected $errors = array();
  protected $adj = 1.75;
  protected $mod = 3;


  /**
   * Decrypts cyphertext.
   *
   * @param $key
   *   String key used for encryption.
   * @param $source
   *   Cyphertext. Text string containing encrypted $source.
   *
   * @return
   *   Plaintext. Text string to be encrypted.
   */
  function decrypt($key, $source) {
    $this->errors = array();

    // Convert key into sequence of numbers
    $fudgefactor = $this->convertKey($key);
    if ($this->errors) {
      return;
    }

    if (empty($source)) {
      // Commented out to prevent errors getting logged for use cases that may
      // have variable encryption/decryption requirements. -RS
      // $this->errors[] = t('No value has been supplied for decryption');
      return;
    }

    $target = NULL;
    $factor2 = 0;

    for ($i = 0; $i < strlen($source); $i++) {
      $char2 = substr($source, $i, 1);

      $num2 = strpos(self::$scramble2, $char2);
      if ($num2 === FALSE) {
        $this->errors[] = t('Source string contains an invalid character (@char)', array('@char' => $char2));
        return;
      }

      $adj = $this->applyFudgeFactor($fudgefactor);
      $factor1 = $factor2 + $adj;
      $num1 = $num2 - round($factor1);
      $num1 = $this->checkRange($num1);
      $factor2 = $factor1 + $num2;

      $char1 = substr(self::$scramble1, $num1, 1);
      $target .= $char1;
    }

    return rtrim($target);
  }

  /**
   * Encrypts plaintext.
   *
   * @param $key
   *   String key used for encryption.
   * @param $source
   *   Plaintext. Text string to be encrypted.
   * @param $sourcelen
   *   Minimum plaintext length.  Plaintext $source which is shorter than
   *   $sourcelen will be padded by appending spaces.
   *
   * @return
   *   Cyphertext. Text string containing encrypted $source.
   */
  function encrypt($key, $source, $sourcelen = 0) {
    $this->errors = array();

    // Convert key into sequence of numbers
    $fudgefactor = $this->convertKey($key);
    if ($this->errors) {
      return;
    }

    if (empty($source)) {
      // Commented out to prevent errors getting logged for use cases that may
      // have variable encryption/decryption requirements. -RS
      // $this->errors[] = t('No value has been supplied for encryption');
      return;
    }

    while (strlen($source) < $sourcelen) {
      $source .= ' ';
    }

    $target = NULL;
    $factor2 = 0;

    for ($i = 0; $i < strlen($source); $i++) {
      $char1 = substr($source, $i, 1);
      echo($char1 . "\n");

      $num1 = strpos(self::$scramble1, $char1);
      if ($num1 === FALSE) {
        $this->errors[] = t('Source string contains an invalid character (@char)', array('@char' => $char1));
        return;
      }
      echo($num1);echo(" <- num1\n");

      $adj = $this->applyFudgeFactor($fudgefactor);
      $factor1 = $factor2 + $adj;
      echo($factor1);echo(" <- factor1\n");
      $num2 = round($factor1) + $num1;
      $num2 = $this->checkRange($num2);
      echo($num2);echo(" <- num2\n");
      $factor2 = $factor1 + $num2;
      $char2 = substr(self::$scramble2, $num2, 1);
      $target .= $char2;
      echo($target . "\n");
    }

    return $target;
  }

  /**
   * Accessor for errors property.
   */
  public function getErrors() {
    return $this->errors;
  }

  /**
   * Mutator for errors property.
   */
  public function setErrors(array $errors) {
    $this->errors = $errors;
  }

  /**
   * Accessor for adj property.
   */
  function getAdjustment() {
    return $this->adj;
  }

  /**
   * Mutator for adj property.
   */
  function setAdjustment($adj) {
    $this->adj = (float) $adj;
  }

  /**
   * Accessor for mod property.
   */
  function getModulus() {
    return $this->mod;
  }

  /**
   * Mutator for mod property.
   */
  function setModulus($mod) {
    $this->mod = (int) abs($mod);
  }

  /**
   * Returns an adjustment value based on the contents of $fudgefactor.
   */
  function applyFudgeFactor(&$fudgefactor) {
    static $alerted = FALSE;

    if (!is_array($fudgefactor)) {
      $fudge = 0;
      if (!$alerted) {
        // Throw an error that makes sense so this stops getting reported.
        $this->errors[] = t('No encryption key was found.');
        drupal_set_message(t('Ubercart cannot find a necessary encryption key.  Refer to the store admin <a href="!url">dashboard</a> to isolate which one.', array('!url' => url('admin/store'))), 'error');

        $alerted = TRUE;
      }
    }
    else {
      $fudge = array_shift($fudgefactor);
    }

    $fudge = $fudge + $this->adj;
    $fudgefactor[] = $fudge;

    if (!empty($this->mod)) {
      if ($fudge % $this->mod == 0) {
        $fudge = $fudge * -1;
      }
    }

    return $fudge;
  }

  /**
   * Checks that $num points to an entry in self::$scramble1.
   */
  function checkRange($num) {
    $num = round($num);
    $limit = strlen(self::$scramble1);

    while ($num >= $limit) {
      $num = $num - $limit;
    }
    while ($num < 0) {
      $num = $num + $limit;
    }

    return $num;
  }

  /**
   * Converts encryption key into an array of numbers.
   *
   * @param $key
   *   Encryption key.
   *
   * @return
   *   Array of integers.
   */
  function convertKey($key) {
    if (empty($key)) {
      // Commented out to prevent errors getting logged for use cases that may
      // have variable encryption/decryption requirements. -RS
      // $this->errors[] = 'No value has been supplied for the encryption key';
      return;
    }

    $array[] = strlen($key);

    $tot = 0;
    for ($i = 0; $i < strlen($key); $i++) {
      $char = substr($key, $i, 1);

      $num = strpos(self::$scramble1, $char);
      if ($num === FALSE) {
        $this->errors[] = "Key contains an invalid character ($char)";
        return;
      }

      $array[] = $num;
      $tot = $tot + $num;
    }

    $array[] = $tot;

    return $array;
  }
}

$enc = new uc_encryption_class();
$b64 = base64_encode('a:9:{s:7:"cc_type";s:10:"Mastercard";s:8:"cc_owner";s:0:"";s:9:"cc_number";s:4:"6441";s:14:"cc_start_month";s:0:"";s:13:"cc_start_year";s:0:"";s:12:"cc_exp_month";s:1:"8";s:11:"cc_exp_year";s:4:"2016";s:8:"cc_issue";s:0:"";s:7:"cc_bank";s:0:"";}');
echo($b64);echo("\n");
$result = $enc->encrypt("111222333dddccccc122456789abcdef", $b64);

var_dump($result);





